### PR TITLE
Add failed initcode to Contract Result

### LINKED ISSUE #4191: Add failed_initcode to contract result
### Problem

As a smart contract developer, I'd like to see results for my contract creates including the initcode.

### Solution

* Add a `contract_result.failed_initcode`
* In `ContractResultServiceImpl`, populate `failed_initcode` from one of the below sources if the transaction is unsuccessful
  * Sidecar bytecode if migration is false
  * `EthereumTransactionBody.ethereum_data.callData`
  * `ContractCreateTransactionBody.initcode`
* Ensure logic doesn't fail anywhere if `ContractBytecode.contractId` is null for these failed entries
* Add `failed_initcode` to the contract result details REST APIs
* Add new field to acceptance tests

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by edwin-greene**
**Description**:
Add Failed Initcode to Contract Result in Importer and REST API.

**Related issue(s)**:

Fixes #4191

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 1]
```diff
@@ -177,6 +177,9 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
             contractResult.setAmount(transactionBody.getInitialBalance());
             contractResult.setFunctionParameters(DomainUtils.toBytes(transactionBody.getConstructorParameters()));
             contractResult.setGasLimit(transactionBody.getGas());
+            if (!recordItem.isSuccessful() && !transactionBody.getInitcode().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (edwin-greene) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Thanks I have implemented this change.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/EthereumTransactionHandler.java [hunk 2]
```diff
@@ -67,6 +68,15 @@ public TransactionType getType() {
         return TransactionType.ETHEREUMTRANSACTION;
     }
 
+    @Override
+    public void updateContractResult(ContractResult contractResult, RecordItem recordItem) {
+        if (!recordItem.isSuccessful()) {
```
[the remark this answers is not in the document]
Comment (edwin-greene) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/EthereumTransactionHandler.java]: Great idea! I have made this change.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java [hunk 3]
```diff
@@ -257,7 +263,8 @@ private void processCreatedContractEntity(RecordItem recordItem, EntityId contra
         entityListener.onEntity(entity);
     }
 
-    private void processSidecarRecords(RecordItem recordItem) {
+    private Optional<ByteString> processSidecarRecords(RecordItem recordItem) {
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java:266]: I don't see the benefit of using `Optional` here. should just return `ByteString`, in addition, `DomainUtils.toBytes(byteString)` returns null when `byteString` is null.

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-rest/viewmodel/contractResultDetailsViewModel.js [hunk 4]
```diff
@@ -64,6 +64,17 @@ class ContractResultDetailsViewModel extends ContractResultViewModel {
...
+    if (this.status === ContractResultDetailsViewModel._FAIL_RESULT) {
+      this.failed_initcode = !_.isEmpty(contractResult.failedInitcode)
+        ? utils.toHexStringNonQuantity(contractResult.failedInitcode)
+        : null;
+      if (_.isNil(this.failed_initcode) && !_.isEmpty(transaction.callData)) {
+        this.failed_initcode = utils.toHexStringNonQuantity(transaction.callData);
+      }
+    } else {
+      this.failed_initcode = null;
+    }
```
Comment (steven-sheehy) [inline hedera-mirror-rest/viewmodel/contractResultDetailsViewModel.js:76]: nit: Could maybe simplify the conditionals:

```suggestion
    if (!_.isEmpty(contractResult.failedInitcode)) {
      this.failed_initcode = utils.toHexStringNonQuantity(contractResult.failedInitcode);
    } else if (this.status === ContractResultDetailsViewModel._FAIL_RESULT && !_.isEmpty(transaction.callData)) {
      this.failed_initcode = utils.toHexStringNonQuantity(transaction.callData);
    }
```

**Review by steven-sheehy — APPROVED** [review]
LGTM

File: hedera-mirror-rest/viewmodel/contractResultDetailsViewModel.js (hunk 4 above)
Comment (edwin-greene) [reply hedera-mirror-rest/viewmodel/contractResultDetailsViewModel.js]: Thanks I have updated the conditional, though I still had one part at the end to set the failed_initcode to null.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java (hunk 3 above)
Comment (edwin-greene) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/domain/ContractResultServiceImpl.java]: Sounds good I have removed the Optional.

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**